### PR TITLE
Dev shahid

### PR DESCRIPTION
**Pull request by ansariDataamanulla**
- fix for libc.so crash and redirection issue after update 
-  fix expected child profile edit issue
- upgrade android sdk 33 to 34
- fix exact alarm permission issue for android

### CODE REVIEW

File: app/components/CountryItem.tsx
Comment (mhdosman) [inline app/components/CountryItem.tsx:1]: @ansariDataamanulla  if these comments are not needed, can you delete them?

File: app/components/homeScreen/DailyReads.tsx
Comment (mhdosman) [inline app/components/homeScreen/DailyReads.tsx:1]: if this code is needed any more, I think you can delete it

**Review by mhdosman — COMMENTED** [review]
all good, you can delete unnecessary commented code

File: .scannerwork/report-task.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+projectKey=bebbo
+serverUrl=http://localhost:9000
+serverVersion=10.6.0.92116
+dashboardUrl=http://localhost:9000/dashboard?id=bebbo
+ceTaskId=5101a772-dfc9-4e3c-b3d1-0292b54796c1
+ceTaskUrl=http://localhost:9000/api/ce/task?id=5101a772-dfc9-4e3c-b3d1-0292b54796c1
```
Comment (Akhror) [inline .scannerwork/report-task.txt:6]: Keep these files out of repository, you can add entire folder to .gitignore file:

.scannerwork/

File: android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmModule.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    public String getName() {
+        return "ExactAlarmModule";
+    }
+
+    @ReactMethod
+    public void requestExactAlarmPermission(Promise promise) {
+        try {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU || Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                AlarmManager alarmManager = (AlarmManager) reactContext.getSystemService(Context.ALARM_SERVICE);
+                if (alarmManager.canScheduleExactAlarms()) {
```
Comment (Akhror) [inline android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmModule.java:34]: // Check if permission is already granted
if (alarmManager != null && alarmManager.canScheduleExactAlarms()) {

File: android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmModule.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+                    reactContext.startActivity(intent); // Open settings to request permission
+                    promise.resolve(false); // Redirecting user to enable the permission
+                }
+            } else {
+                promise.resolve(true); // Permission not needed for versions below Android 12
+            }
+        } catch (Exception e) {
+            promise.reject("ERROR", e.getMessage()); // Handle exception if permission check fails
+        }
```
Comment (Akhror) [inline android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmModule.java:48]: Catch specific exceptions first.

catch (SecurityException e) {
            promise.reject("SECURITY_EXCEPTION", "Failed to request exact alarm permission: " + e.getMessage());
        } catch (Exception e) {
            promise.reject("ERROR", "Unexpected error: " + e.getMessage());
        }

File: android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmPackage.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        return Collections.emptyList();
+    }
+
+    @Override
+    public List<NativeModule> createNativeModules(ReactApplicationContext reactContext) {
+        List<NativeModule> modules = new ArrayList<>();
+        modules.add(new ExactAlarmModule(reactContext));
+        return modules;
+    }
+}
```
Comment (Akhror) [inline android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmPackage.java:24]: Add some comments to the class and methods.

File: app/components/FavActivities.tsx [hunk 5]
```diff
@@ -59,7 +59,9 @@ const FavActivities = (): any => {
     state.childData.childDataSet.favoritegames
   );
   const [favGamesToShow, setfavGamesToShow] = useState([]);
-  const activityTaxonomyId = activeChild?.taxonomyData.prematureTaxonomyId != null && activeChild?.taxonomyData.prematureTaxonomyId != undefined && activeChild?.taxonomyData.prematureTaxonomyId != "" ? activeChild?.taxonomyData.prematureTaxonomyId : activeChild?.taxonomyData.id;
+
+  const activityTaxonomyId = activeChild?.taxonomyData?.prematureTaxonomyId ?? activeChild?.taxonomyData?.id;
```
Comment (Akhror) [inline app/components/FavActivities.tsx:63]: I would name it as 
"childActivityTaxonomyId" or "taxonomyIdForActiveChild"

File: app/components/homeScreen/DailyReads.tsx [hunk 6]
```diff
@@ -307,13 +314,20 @@ const DailyReads = (): any => {
...
     }
-  }, [activeChild?.uuid, activityTaxonomyId]);
+  }, [activeChild?.uuid, activityTaxonomyId,fetchAgain]);
   const keyExtractor = useCallback((item: any) => item?.id, []);
 
   const renderItem = useCallback((item: any, index: any, isAdvice: boolean) => {
     return <RenderDailyReadItem item={item} index={index} isAdvice={isAdvice} />;
   }, [favoriteAdvices, favoriteGames]);
 
+  const renderEmptyList = () => {
```
Comment (Akhror) [inline app/components/homeScreen/DailyReads.tsx:324]: Function does not render anything when the list is empty, it just returns null. Maybe we should rename it to handleEmptyList

File: app/components/homeScreen/DailyReads.tsx [hunk 7]
```diff
@@ -307,13 +314,20 @@ const DailyReads = (): any => {
...
+      if(dataToShowInList?.length == 0 && !fetchAgain){
```
Comment (Akhror) [inline app/components/homeScreen/DailyReads.tsx:325]: Use strict equality to improve type safety: ===

File: app/screens/localization/LanguageSelection.tsx [hunk 8]
```diff
@@ -119,15 +120,15 @@ const LanguageSelection = ({ route, navigation }: Props): any => {
         } else {
 
           if (route.params != undefined) {
-            newCountryId = route.params.country.countryId;
+            newCountryId = route.params.country?.countryId;
```
Comment (Akhror) [inline app/screens/localization/LanguageSelection.tsx:123]: Can params be null? If yes should we add "?"

File: app/screens/localization/LanguageSelection.tsx [hunk 9]
```diff
@@ -229,42 +230,49 @@ const LanguageSelection = ({ route, navigation }: Props): any => {
...
+      navigation.navigate("LoadingScreen", params );
     }
   }
   const goToConfirmationScreen = (): any => {
-    let newLanguage: any = null
+    let newLanguage: any = null;
     if (language?.locale != undefined) {
-      newLanguage = language
+      newLanguage = language;
+    } else if (Array.isArray(language) && language.length > 0) {
```
Comment (Akhror) [inline app/screens/localization/LanguageSelection.tsx:262]: language?.length

File: app/services/ScrollingButtonMenu.js [hunk 10]
```diff
@@ -153,16 +153,16 @@ export default class ScrollingButtonMenu extends React.Component {
                     keyboardShouldPersistTaps={keyboardShouldPersistTaps}
                 >
                     {
-                        items.map((route, i) => (
+                      (items?.length > 0) &&  items.map((route, i) => (
```
Comment (Akhror) [inline app/services/ScrollingButtonMenu.js:156]: if there is a possibility that items can be null then we add "?"
items?.map(...)

File: app/services/childCRUD.ts [hunk 11]
```diff
@@ -560,7 +565,8 @@ export const updateActiveChild = (child: any, key: any, value: any, dispatch: an
 }
 export const getAllConfigData = async (dispatch: any): Promise<any> => {
   const allJsonDatanew = await dataRealmCommon.getData<ConfigSettingsEntity>(ConfigSettingsSchema);
-  allJsonDatanew.removeAllListeners();
+  console.log("=====2",allJsonDatanew)
```
Comment (Akhror) [inline app/services/childCRUD.ts:568]: Make sure these kind of logs do not go to production.

File: app/services/exactAlarmService.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+const { ExactAlarmModule } = NativeModules;
+
+export const requestExactAlarmPermission = async (): Promise<boolean> => {
+  try {
+    // Request permission from the native module
+    const permissionGranted = await ExactAlarmModule.requestExactAlarmPermission();
+    // Return true if permission is granted, otherwise false
+    return permissionGranted;
+  } catch (error: any) {
+    console.log('Error requesting exact alarm permission:', error.message || error);
```
Comment (Akhror) [inline app/services/exactAlarmService.ts:12]: console.error for error messages will distinguish errors from other log messages.

File: app/services/notificationService.ts [hunk 13]
```diff
@@ -140,8 +140,8 @@ export const getCDGWNotisForChild = (childTaxonomyData: any, child: any, prematu
   // noti.push(//at the begginning of the period
   //if> 3months then add noti
   //check gap between days_to and days_from
-  const childBirthDatePlanned = child?.taxonomyData.prematureTaxonomyId != null && child?.taxonomyData.prematureTaxonomyId != "" && child?.taxonomyData.prematureTaxonomyId != undefined ? child.plannedTermDate : child.birthDate;
-  const activityTaxonomyId = child?.taxonomyData.prematureTaxonomyId != null && child?.taxonomyData.prematureTaxonomyId != undefined && child?.taxonomyData.prematureTaxonomyId != "" ? child?.taxonomyData.prematureTaxonomyId : child?.taxonomyData.id;
+  const childBirthDatePlanned = child?.taxonomyData?.prematureTaxonomyId ?? child?.birthDate;
```
Comment (Akhror) [inline app/services/notificationService.ts:143]: Updated code is not the same as it was before, please make sure it is intentional.

File: app/services/notificationService.ts [hunk 14]
```diff
@@ -307,14 +307,14 @@ export const getNextChildNotification = (gwperiodid: any, vcperiodid: any, hcper
...
-  const childBirthDatePlanned = child?.taxonomyData.prematureTaxonomyId != null && child?.taxonomyData.prematureTaxonomyId != "" && child?.taxonomyData.prematureTaxonomyId != undefined ? child.plannedTermDate : child.birthDate;
+  const childBirthDatePlanned = child?.taxonomyData?.prematureTaxonomyId ?? child.birthDate;
   const childAgeInDaysForCD = getCurrentChildAgeInDays(
     DateTime.fromJSDate(new Date(childBirthDatePlanned)).toMillis(),
   );
   const childAgeObj = childAge.sort(
     (a: any, b: any) => a.days_from - b.days_from,
   );
-  const activityTaxonomyId = child?.taxonomyData.prematureTaxonomyId != null && child?.taxonomyData.prematureTaxonomyId != undefined && child?.taxonomyData.prematureTaxonomyId != "" ? child?.taxonomyData.prematureTaxonomyId : child?.taxonomyData.id;
+  const activityTaxonomyId = child?.taxonomyData?.prematureTaxonomyId != null && child?.taxonomyData?.prematureTaxonomyId != undefined && child?.taxonomyData?.prematureTaxonomyId != "" ? child?.taxonomyData?.prematureTaxonomyId : child?.taxonomyData.id;
```
Comment (Akhror) [inline app/services/notificationService.ts:317]: Should we simplify this one as well?

File: app/services/notificationService.ts [hunk 15]
```diff
@@ -326,8 +326,8 @@ export const getNextChildNotification = (gwperiodid: any, vcperiodid: any, hcper
...
     //find next period and calculate for it and return for gw cd noti
     for (let i = lastchildgwperiodIndex + 1; i <= currentchildgwperiodIndex; i++) {
-      const childDaysTo = child?.taxonomyData.prematureTaxonomyId != null && child?.taxonomyData.prematureTaxonomyId != undefined && child?.taxonomyData.prematureTaxonomyId != "" ? prematurechildgwperiod.days_to : childAgeObj[i].days_to;
-      const childDaysFrom = child?.taxonomyData.prematureTaxonomyId != null && child?.taxonomyData.prematureTaxonomyId != undefined && child?.taxonomyData.prematureTaxonomyId != "" ? prematurechildgwperiod.days_from : childAgeObj[i].days_from;
+      const childDaysTo = child?.taxonomyData?.prematureTaxonomyId != null && child?.taxonomyData?.prematureTaxonomyId != undefined && child?.taxonomyData?.prematureTaxonomyId != "" ? prematurechildgwperiod.days_to : childAgeObj[i].days_to;
+      const childDaysFrom = child?.taxonomyData?.prematureTaxonomyId != null && child?.taxonomyData?.prematureTaxonomyId != undefined && child?.taxonomyData?.prematureTaxonomyId != "" ? prematurechildgwperiod.days_from : childAgeObj[i].days_from;
```
Comment (Akhror) [inline app/services/notificationService.ts:330]: Should we simplify this code?

File: app/components/homeScreen/DailyReads.tsx
Comment (ansariDataamanulla) [reply app/components/homeScreen/DailyReads.tsx:1]: we commented this code for now to check isChildPremature or not we need this code in future

File: android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmModule.java (hunk 2 above)
Comment (ansariDataamanulla) [reply android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmModule.java:34]: changed as you suggested

File: android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmModule.java (hunk 3 above)
Comment (ansariDataamanulla) [reply android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmModule.java:48]: handled specific exception

File: android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmPackage.java (hunk 4 above)
Comment (ansariDataamanulla) [reply android/app/src/main/java/org/unicef/ecar/bebbo/ExactAlarmPackage.java:24]: added appropriate comment

File: app/components/homeScreen/DailyReads.tsx (hunk 6 above)
Comment (ansariDataamanulla) [reply app/components/homeScreen/DailyReads.tsx:324]: change the name

File: app/components/homeScreen/DailyReads.tsx (hunk 7 above)
Comment (ansariDataamanulla) [reply app/components/homeScreen/DailyReads.tsx:325]: added strict equality

File: app/screens/localization/LanguageSelection.tsx (hunk 8 above)
Comment (ansariDataamanulla) [reply app/screens/localization/LanguageSelection.tsx:123]: added the ? for handle if params getting null

File: app/screens/localization/LanguageSelection.tsx (hunk 9 above)
Comment (ansariDataamanulla) [reply app/screens/localization/LanguageSelection.tsx:262]: in this we have 2 condition if we get as object and as array we handled both ways

File: app/services/childCRUD.ts (hunk 11 above)
Comment (ansariDataamanulla) [reply app/services/childCRUD.ts:568]: done

File: app/services/exactAlarmService.ts (hunk 12 above)
Comment (ansariDataamanulla) [reply app/services/exactAlarmService.ts:12]: changes done as suggested

File: app/services/notificationService.ts (hunk 13 above)
Comment (ansariDataamanulla) [reply app/services/notificationService.ts:143]: yes it's intentional it's optimise condition using nullish coalescing operator

File: app/services/notificationService.ts (hunk 14 above)
Comment (ansariDataamanulla) [reply app/services/notificationService.ts:317]: optimise the condition

File: app/services/notificationService.ts (hunk 15 above)
Comment (ansariDataamanulla) [reply app/services/notificationService.ts:330]: done

**Review by Akhror — APPROVED** [review]

**Pull request merged**